pkg/kubelet/status/manager.go

```diff
@@ -392,6 +405,94 @@ func (m *manager) needsUpdate(uid types.UID, status versionedPodStatus) bool {
+		return false
+	}
+	// If the pod is a static pod, we should check its mirror pod, because only status in mirror pod is meaningful to us.
+	if kubepod.IsStaticPod(pod) {
+		mirrorPod, ok := m.podManager.GetMirrorPodByPod(pod)
+		if !ok {
+			glog.V(4).Infof("Static pod %q has no corresponding mirror pod, no need to reconcile", format.Pod(pod))
+			return false
+		}
+		pod = mirrorPod
+	}
+
+	// Normalize the status from pod manager, in fact the status should have been normalized (by apiserver), that's
+	// why we have to normalize the cached status before comparing here. However, in case that the implementation on
+	// apiserver may be changed someday, also normalize the status passed in here.
+	// TODO: (random-liu) When timestamp is made consistent on client and server side, remove all the following
+	// normalization.
+	normalizedStatus := normalizeStatus(pod.Status)
+	// Normalize the cached status
+	normalizedCachedStatus := normalizeStatus(status)
+	if isStatusEqual(&normalizedStatus, &normalizedCachedStatus) {
+		// If the status from the source is the same with the cached status,
+		// reconcile is not needed. Just return.
+		return false
+	}
+	glog.V(3).Infof("Pod status is inconsistent with cached status, a reconcile should be triggered: %+v", util.ObjectDiff(normalizedStatus, normalizedCachedStatus))
+
+	return true
+}
+
+// We add this function, because apiserver only supports *RFC3339* now, which means that the timestamp returned by
+// apiserver has no nanosecond infromation. However, the timestamp returned by unversioned.Now() contains nanosecond,
+// so when we do comparation between status from apiserver and cached status, isStatusEqual() will always return false.
+// There is related issue #15262 and PR #15263 about this.
+// In fact, the best way to solve this is to do it on api side. However for now, we normalize the status locally in
+// kubelet temporarily.
+// TODO: (random-liu) Remove this after apiserver makes it consistent on api side.
+func normalizeStatus(cachedStatus api.PodStatus) api.PodStatus {
+	// Make sure we don't modify the cached status
+	clone, err := api.Scheme.DeepCopy(&cachedStatus)
+	if err != nil {
+		glog.Errorf("Failed to clone status %+v: %v", cachedStatus, err)
+		// If failed to clone the status, just return the cached status. It only causes
+		// isStatusEqual() return false unexpectedly, which is not a big deal if it
+		// doesn't happen usually.
+		return cachedStatus
+	}
+	status := *clone.(*api.PodStatus)
+	normalizeTimeStamp := func(t *unversioned.Time) {
+		*t = t.Rfc3339Copy()
+	}
+	normalizeContainerState := func(c *api.ContainerState) {
+		if c.Running != nil {
+			normalizeTimeStamp(&c.Running.StartedAt)
+		}
+		if c.Terminated != nil {
+			normalizeTimeStamp(&c.Terminated.StartedAt)
+			normalizeTimeStamp(&c.Terminated.FinishedAt)
+		}
+	}
+
+	if status.StartTime != nil {
+		normalizeTimeStamp(status.StartTime)
+	}
+	for i := range status.Conditions {
```

Comment:
If implementing like this, the `condition` will be a copy, right? I need to modify the original object, so I use index instead. Is there any other ways to do that?